tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,434 @@
+                    if (name == "s")
+                    {
+                        // SamplingPriorityToInt32(string)
+                        samplingPriority = SamplingPriorityToInt32(value);
+                    }
+                    else if (name == "o")
+                    {
+                        origin = value;
+                    }
+                    else if (name.StartsWith("t.", StringComparison.Ordinal))
+                    {
+                        propagatedTagsBuilder.Append("_dd.p.").Append(name.Substring(2)).Append('=').Append(value).Append(',');
+                    }
 #endif
 
+                    // skip past the semicolon
+                    startIndex = endIndex + 2;
+                }
+
+                string? propagatedTags;
+
+                if (propagatedTagsBuilder.Length > 0)
+                {
+                    if (propagatedTagsBuilder[propagatedTagsBuilder.Length - 1] == ',')
+                    {
+                        propagatedTagsBuilder.Length--;
+                    }
+
+                    propagatedTags = StringBuilderCache.GetStringAndRelease(propagatedTagsBuilder);
+                }
+                else
+                {
+                    propagatedTags = null;
+                }
+
+                traceState = new W3CTraceState(samplingPriority, origin, propagatedTags);
                 return true;
             }
+            finally
+            {
+                StringBuilderCache.Release(propagatedTagsBuilder);
+            }
+        }
+
+        [return: NotNullIfNotNull("samplingPriority")]
+        private static string? SamplingPriorityToString(int? samplingPriority)
+        {
+            return samplingPriority switch
+                   {
+                       2 => "2",
+                       1 => "1",
+                       0 => "0",
+                       -1 => "-1",
+                       null => null,
+                       not null => samplingPriority.Value.ToString(CultureInfo.InvariantCulture)
+                   };
+        }
 
-            return false;
+        private static int? SamplingPriorityToInt32(string? samplingPriority)
+        {
+            return samplingPriority switch
+                   {
+                       "2" => 2,
+                       "1" => 1,
+                       "0" => 0,
+                       "-1" => 1,
+                       null => null,
+                       not null => int.TryParse(samplingPriority, out var result) ? result : null
+                   };
         }
 
-        private static bool IsValidTraceId([NotNullWhen(true)] string? traceId)
+#if NETCOREAPP
+        private static int? SamplingPriorityToInt32(ReadOnlySpan<char> samplingPriority)
         {
-            if (string.IsNullOrEmpty(traceId))
+            if (samplingPriority.Equals("2", StringComparison.Ordinal))
```

Comment:
Why `Equals` here, but `SequenceEquals` in the previous func? What's the difference?

Reply:
`ReadOnlySpan<T>.Equals(object)` always throws a `NotSupportedException` ([docs](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.equals?view=net-7.0)), and the equality operator compares the internal values (memory point and length), so it doesn't work for string equality.

The solution I found to compare `ReadOnlySpan<char>` as strings was to use either of these extension methods:
```
ReadOnlySpan<T>.Equals(ReadOnlySpan<T>, StringComparison)
ReadOnlySpan<T>.SequenceEquals(ReadOnlySpan<T>)
```

Looks at the docs now, `SequenceEquals` will call `char.Equals` on each `char`, so `Equals` with `StringComparison` is probably better. I'll fix these, thanks.

Reply:
Done.